pkg/server/server.go

```diff
@@ -141,6 +145,90 @@ var _ agent.AgentServiceServer = &ProxyServer{}
 
 var _ client.ProxyServiceServer = &ProxyServer{}
 
+func genContext(proxyStrategies []ProxyStrategy, reqHost string) context.Context {
+	ctx := context.Background()
+	for _, ps := range proxyStrategies {
+		switch ps {
+		case ProxyStrategyDestHost:
+			addr := util.RemovePortFromHost(reqHost)
+			ctx = context.WithValue(ctx, "destHost", addr)
+		}
+	}
+	return ctx
+}
+
+func (s *ProxyServer) getBackend(reqHost string) (Backend, error) {
+	ctx := genContext(s.proxyStrategies, reqHost)
+	for _, bm := range s.BackendManagers {
+		be, err := bm.Backend(ctx)
+		if err == nil {
+			return be, nil
+		}
+		if ignoreNotFound(err) != nil {
+			// if can't find a backend through current BackendManager, move on
+			// to the next one
+			return nil, err
+		}
+	}
+	return nil, &ErrNotFound{}
+}
+
+func (s *ProxyServer) addBackend(agentID string, conn agent.AgentService_ConnectServer) (backend Backend) {
+	for i := 0; i < len(s.BackendManagers); i++ {
+		switch s.BackendManagers[i].(type) {
+		case *DestHostBackendManager:
+			agentAddress, err := getAgentAddress(conn)
+			if err != nil {
+				klog.ErrorS(err, "fail to get the agent Host", "agentID", agentID)
+				break
+			}
+			for _, ipv4 := range agentAddress.IPv4 {
+				klog.V(5).InfoS("Add the agent to DestHostBackendManager", "agent address", ipv4)
+				s.BackendManagers[i].AddBackend(ipv4, conn)
+			}
+			for _, ipv6 := range agentAddress.IPv6 {
+				klog.V(5).InfoS("Add the agent to DestHostBackendManager", "agent address", ipv6)
+				s.BackendManagers[i].AddBackend(ipv6, conn)
+			}
+			for _, host := range agentAddress.Host {
+				klog.V(5).InfoS("Add the agent to DestHostBackendManager", "agent address", host)
+				s.BackendManagers[i].AddBackend(host, conn)
+			}
+		default:
+			klog.V(5).InfoS("Add the agent to DefaultBackendManager", "agentID", agentID)
+			backend = s.BackendManagers[i].AddBackend(agentID, conn)
```

Comment:
agentID is the universal ID for the agent and shouldn't be conflated with the other IDs. As an example of default backend manager should probably just ignore anything which is not the universal ID. I would also specify and track the "type" of the id into the backend manager. That way its easy to write a backend manager which only care about something like exact IP match and will not confuse an IP with anything else.

Reply:
To track the identifier type in BackendManager, I added a new field `idTypes` to the `DefaultBackendStorage`, which contains a list of valid identifier types. When calling `AddBackend`, the Identifiertype will also need to be passed in to the function, if the identifier type is not included in the `idTypes`, the BackendStorage will not add it to the backend list.